src/Hpack/Config.hs

```diff
@@ -150,11 +153,14 @@ instance FromJSON (CaptureUnknownFields FlagSection) where
 
 getUnknownFields :: forall a. HasFieldNames a => Value -> Proxy a -> [FieldName]
 getUnknownFields v _ = case v of
-  Object o -> unknown
+  Object o -> ignoreUnderscored unknown
     where
       unknown = keys \\ fields
       keys = map T.unpack (HashMap.keys o)
       fields = fieldNames (Proxy :: Proxy a)
+      ignoreUnderscored
+        | warnUnderscoredUnknownFields (Proxy :: Proxy a) = id
+        | otherwise = filter (not . isPrefixOf "_")
```

Comment:
Ah, I see, it's because we treat unknown fields from top-level conditionals the same way as unknown top-level fields.  In that case let's go with your solution.

Given that there is only one more thing I would ask.  Can we rename `warnUnderscoredUnknownFields` to `ignoreUnderscoredUnknownFields` inverse the bools.  That code is not really concerned with warnings, but rather with collecting unknown field names.